_posts/2021-04-28-#17331.md

```diff
@@ -0,0 +1,73 @@
+---
+layout: pr
+date: 2021-04-28
+title: "Use effective values throughout coin selection"
+pr: 17331
+authors: [achow101]
+components: ["wallet"]
+host: Xekyo
+status: upcoming
+commit: 4ac1add
+---
+
+## Notes
+
+- Wallets maintain their balance as a pool of Unspent Transaction Outputs
+  (UTXOs).
+
+- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
+  pool to fund a transaction.
+
+- What are the goals of coin selection? Economics, privacy, reliability of
+  inputs/conflicts
+
+- Why are we trying to not spend unconfirmed inputs?
+
+- When a new transaction is built, the target feerate is provided as an input.
+
+- A UTXOs _effective value_ is its value reduced by its cost to be spent at a
+  given feerate.
+
+- Using _effective values_ in coin selection means that the cost of spending the
+  UTXO has already been accounted for when a UTXO gets picked.
+
+- The knapsacksolver tries to combine input sets by randomly selecting or
+  skipping inputs until it has sufficient funds for the transaction. When it
+  finds a solution, it deselects the last input and tries to find one with a
+  smaller change output. It repeats this random drawing 1000 times.
+
+- Branch and Bound deterministically searches the complete combination space. It
+  walks a binary search tree where each branch is the inclusion or exclusion of
+  a UTXO. It explores the inclusion branch first, and backtracks whenever a
+  subtree cannot yield a solution. It returns the first input set that is an
+  exact match for the funding requirement of the transaction.
+
+- OutputGroups collect all UTXOs that were sent to the same public key.
+
```

Comment:
Maybe add a bullet or two about what this PR is changing.

Reply:
Added a big bullet :)